src/python/alog/protocols.py

```diff
@@ -0,0 +1,576 @@
+################################################################################
+# MIT License
+#
+# Copyright (c) 2021 IBM
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+################################################################################
+"""Protocols to assist IDEs with code completion for the additional methods which
+Alchemy Logger overrides on the logging.Logger instances.
+"""
+
+import sys
+from logging import Filter, LogRecord
+from typing import (
+    TYPE_CHECKING,
+    Callable,
+    ClassVar,
+    List,
+    Mapping,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+)
+
+from .alog import _Level
+
+if sys.version_info < (3, 8):
+    from typing_extensions import Protocol
+else:
+    from typing import Protocol
+
+if TYPE_CHECKING:
+    from logging import (
+        Handler,
+        Logger,
+        Manager,
+        RootLogger,
+        _ArgsType,
+        _ExcInfoType,
+        _SysExcInfoType,
+    )
+
+    from typing_extensions import Self
```

Comment:
Will this fail for >= `3.8`?

Reply:
I think since it's behind `if TYPE_CHECKING`, there is no realistic circumstance where it would be hit by anything besides `mypi`. That said, if there were some library that used `typing.TYPE_CHECKING = True` and imported `alog` without having `typing_extensions` installed from elsewhere, I think this _would_ fail.